test/replication-luatest/bootstrap_strategy_test.lua

```diff
@@ -774,3 +779,87 @@ g_supervised.test_no_bootstrap_without_replication = function(cg)
+            -- successfully bootstraps the database if no
+            -- upstreams are configured.
+        },
+        env = {
+            ['TARANTOOL_RUN_BEFORE_BOX_CFG'] =
+                'box.ctl.make_bootstrap_leader()',
+        },
+    }
+end)
+
+-- If the given instance is chosen as a bootstrap leader before a
+-- first box.cfg() call, it is expected to successfully bootstrap
+-- the database even if there are no upstreams configured.
+--
+-- This logic is added in gh-10858.
+g_supervised.test_early_leader_singleton = function(cg)
+    cg.server1:start()
+
+    t.assert_equals(cg.server1:get_instance_id(), 1,
+                    'Server 1 is the bootstrap leader')
+
+    cg.server1:exec(function()
+        local tup = box.space._schema:get{'bootstrap_leader_uuid'}
+        t.assert(tup ~= nil, 'Bootstrap leader uuid is persisted')
+        t.assert_equals(tup[2], box.info.uuid,
+                        'Bootstrap leader uuid is correct')
+    end)
+end
+
+g_supervised.before_test('test_early_leader_with_replica', function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.box_cfg = {
+        bootstrap_strategy = 'supervised',
+        replication = {
+            server.build_listen_uri('server1', cg.replica_set.id),
+            server.build_listen_uri('server2', cg.replica_set.id),
+            -- Don't start this third server to verify that once
+            -- a bootstrap leader is found, a replica doesn't wait
+            -- anything else to register in the replicaset.
+            server.build_listen_uri('server3', cg.replica_set.id),
+        },
+        replication_timeout = 0.1,
+        -- This way we verify that when the replica connects to
+        -- the bootstrap leader, it correctly determines that it
+        -- is the bootstrap leader and doesn't wait anything else.
+        replication_connect_timeout = 1000,
+    }
+    for i = 1, 2 do
+        local alias = 'server' .. i
+        cg[alias] = cg.replica_set:build_and_add_server{
+            alias = alias,
+            box_cfg = cg.box_cfg,
+        }
+    end
+
+    cg.server2.env.TARANTOOL_RUN_BEFORE_BOX_CFG =
+        'box.ctl.make_bootstrap_leader()'
+end)
+
+-- If the given instance is chosen as a bootstrap leader before a
+-- first box.cfg() call, it is expected that it is successfully
+-- bootstrapped and bootstraps a replica.
+--
+-- This logic is added in gh-10858.
+g_supervised.test_early_leader_with_replica = function(cg)
+    cg.replica_set:start()
+    t.assert_equals(cg.server2:get_instance_id(), 1,
+                    'Server 2 is the bootstrap leader')
+    cg.server2:exec(function()
+        local tup = box.space._schema:get{'bootstrap_leader_uuid'}
+        t.assert(tup ~= nil, 'Bootstrap leader uuid is persisted')
+        t.assert_equals(tup[2], box.info.uuid,
+                        'Bootstrap leader uuid is correct')
+    end)
+    t.helpers.retrying({}, cg.server1.assert_follows_upstream, cg.server1, 1)
+end
```

Comment:
I suppose, we also need a test, where several `box.ctl.make_bootstrap_leader` are called before box.cfg. But let's firstly discuss, what the expected behavior in such case

Reply:
Judging from Sergey's answer in the above thread, it seems that we expect `ER_REPLICASET_UUID_MISMATCH` for now.

I think that the best course is to verify this behavior and link an issue with proposal to add the safety checks. I'll create the issue.

@Serpentian @sergepetrenko Are you OK?

Reply:
I'm ok

Reply:
Added `test_early_leader_several_leaders`, please, take a look.

I also decided to refuse to look into applier state in `replicaset_find_join_master_supervised` if the given instance is assigned as a bootstrap leader, because even in local testing it catches ~3% of situations and adds instability to testing. These 3% (or even less in real network conditions) don't look valuable for a user.

If we need the checks of this kind, we should guarantee some wait interval, I think.

Now the function looks as follows.

```c
/**
 * Out of all the replicas find the one which was promoted via
 * box.ctl.make_bootstrap_leader()
 */
static struct replica *
replicaset_find_join_master_supervised(void)
{
        /*
         * If we're known as a bootstrap leader, just bootstrap
         * the database.
         *
         * Unlike the 'auto' bootstrap strategy, the 'supervised'
         * one has no safety checks with waiting for ballots from
         * other replicaset members.
         *
         * NB: It is possible to run the cycle below first and
         * verify other members status known at this moment.
         * However, since we don't wait for connecting to all the
         * instances, it has a low chance to catch a problem.
         * As result, it would just make the result of calling
         * this function less predictable without visible gains
         * for a user.
         */
        if (is_supervised_bootstrap_leader)
                return NULL;

        struct replica *leader = NULL;
        replicaset_foreach(replica) {
                struct applier *applier = replica->applier;
                if (applier == NULL)
                        continue;
                if (applier_is_bootstrap_leader(applier))
                        leader = replica;
        }
        if (leader == NULL) {
                tnt_raise(ClientError, ER_CFG, "bootstrap_strategy",
                          "failed to connect to the bootstrap leader");
        }
        return leader;
}
```